### PR TITLE
[d3d9, dxso] Clamp Dref to [0.0, 1.0] if the texture is emulated UNORM

### PR DESCRIPTION
**Pull request by DadSchoorse**
Fixes: https://github.com/doitsujin/dxvk/issues/2693
Fixes: https://github.com/doitsujin/dxvk/issues/2361
Fixes: https://github.com/doitsujin/dxvk/issues/1814
fixes shadows in https://github.com/doitsujin/dxvk/issues/2029 (no idea if the line is still an issue on amd)

### CODE REVIEW

File: src/d3d9/d3d9_common_texture.cpp [hunk 1]
```diff
@@ -40,6 +40,8 @@ namespace dxvk {
 
     m_mapMode        = DetermineMapMode();
     m_shadow         = DetermineShadowState();
+    m_upgradedToD32f = ConvertFormatUnfixed(m_desc.Format).FormatColor != VK_FORMAT_D32_SFLOAT_S8_UINT &&
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_common_texture.cpp:43]: Just set this after image creation and check if the VkFormat matches the desc format.

File: src/d3d9/d3d9_device.cpp [hunk 2]
```diff
@@ -3791,6 +3791,9 @@ namespace dxvk {
         m_dirtySamplerStates |= 1u << StateSampler;
       }
 
+      m_drefClamp &= ~(1u << StateSampler);
+      m_drefClamp |= newTexture->IsUpgradedToD32f() << StateSampler;
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:3795]: Shifting a bool isn't going to work for samplers >= 8.

Just write this the same as the others.

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_spec_constants.h [hunk 3]
```diff
@@ -28,6 +28,7 @@ namespace dxvk {
     SpecPixelShaderBools,   // 16 bools                       | Bits: 16
 
     SpecFetch4,             // 1 bit for 16 PS samplers       | Bits: 16
+    SpecDrefClamp,          // 1 bit for 20 VS + PS samplers  | Bits: 20
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_spec_constants.h:31]: Veeeery minor nit, but I'd move this before Fetch4 because its more common.

(Same for bit definition below.)

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_spec_constants.h [hunk 4]
```diff
@@ -62,7 +63,8 @@ namespace dxvk {
       { 3, 0,  16 }, // VertexShaderBools
       { 3, 16, 16 }, // PixelShaderBools
 
-      { 4, 0,  16 }, // Fetch4
+      { 4, 0,  20 }, // DrefClamp
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_spec_constants.h:66]: Actually, just put this in the same as FETCH4.

Its not possible to do projections in vs, so it can fit.

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_common_texture.cpp (hunk 1 above)
Comment (Joshua-Ashton) [reply src/d3d9/d3d9_common_texture.cpp:43]: If desc.format == D24 && image->info->format == VK_Format_d32

Comment (DadSchoorse) [reply src/d3d9/d3d9_common_texture.cpp:43]: well but if a driver doesn't support D16 and D24, that's not good enough.

Comment (Joshua-Ashton) [reply src/d3d9/d3d9_common_texture.cpp:43]: Sure.

**Pull request merged**